sys/include/net/sock/ip.h

```diff
@@ -0,0 +1,204 @@
+ * @param[in] proto     Protocol to use in the raw IPv4/IPv6 sock object
+ *                      (the `protocol` header field in IPv4 and the `next_header`
+ *                      field in IPv6).
+ * @param[in] flags     Flags for the sock object. See also @ref net_sock_flags.
+ *                      May be 0.
+ *
+ * @return  0 on success.
+ * @return  -EADDRINUSE, if `local != NULL` and the stack reports that @p local
+ *          is already used elsewhere
+ * @return  -EAFNOSUPPORT, if `local != NULL` or `remote != NULL` and
+ *          sock_ip_ep_t::family of @p local or @p remote is not supported.
+ * @return  -EINVAL, if `proto` is not supported or if sock_ip_ep_t::netif of
+ *          @p local or @p remote is not a valid interface or contradict each
+ *          other (i.e. `(local->netif != remote->netif) &&
+ *          ((local->netif != SOCK_ADDR_ANY_NETIF) ||
+ *          (remote->netif != SOCK_ADDR_ANY_NETIF))` if neither is `NULL`).
+ * @return  -ENOMEM, if the stack can't provide enough resources for `sock` to
+ *          be created.
+ * @return  -EPROTONOSUPPORT, if `local != NULL` or `remote != NULL` and
+ *          proto is not supported by sock_ip_ep_t::family of @p local or @p
+ *          remote.
+ */
+int sock_ip_create(sock_ip_t *sock, const sock_ip_ep_t *local,
+                   const sock_ip_ep_t *remote, uint8_t proto, uint16_t flags);
+
+/**
+ * @brief   Closes a raw IPv4/IPv6 sock object
+ *
+ * @pre `(sock != NULL)`
+ *
+ * @param[in] sock  A raw IPv4/IPv6 sock object.
+ */
+void sock_ip_close(sock_ip_t *sock);
+
+/**
+ * @brief   Gets the local end point of a raw IPv4/IPv6 sock object
+ *
+ * @pre `(sock != NULL) && (ep != NULL)`
+ *
+ * @param[in] sock  A raw IPv4/IPv6 sock object.
+ * @param[out] ep   The local end point.
+ *
+ * @return  0 on success.
+ * @return  -EADDRNOTAVAIL, when @p sock has no end point bound to it.
+ */
+int sock_ip_get_local(sock_ip_t *sock, sock_ip_ep_t *ep);
+
+/**
+ * @brief   Gets the remote end point of a UDP sock object
+ *
+ * @pre `(sock != NULL) && (ep != NULL)`
+ *
+ * @param[in] sock  A UDP sock object.
+ * @param[out] ep   The remote end point.
+ *
+ * @return  0 on success.
+ * @return  -ENOTCONN, when @p sock has no remote end point bound to it.
+ */
+int sock_ip_get_remote(sock_ip_t *sock, sock_ip_ep_t *ep);
+
+/**
+ * @brief   Receives a message over IPv4/IPv6 from remote end point
+ *
+ * @pre `(sock != NULL) && (data != NULL) && (max_len > 0)`
+ *
+ * @param[in] sock      A raw IPv4/IPv6 sock object.
+ * @param[out] data     Pointer where the received data should be stored.
+ * @param[in] max_len   Maximum space available at @p data.
+ *                      If received data exceeds @p max_len the data is
+ *                      truncated and the remaining data can be retrieved
+ *                      later on.
+ * @param[in] timeout   Timeout for receive in microseconds.
+ *                      This value can be ignored (no timeout) if the
```

Comment:
I find it weird for a API to require a module. What if the implementation comes with its own timer implementation?